### PR TITLE
java11_tomcat9 image PR

### PR DESCRIPTION
**Pull request by naveen-armedia**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1 +1,27 @@
+FROM 345280441424.dkr.ecr.ap-south-1.amazonaws.com/ark_base_java11:latest
+RUN useradd -r -g root tomcat
```
Comment (fabricetriboix) [inline Dockerfile]: The `tomcat` user should not be in the `root` group, that will give it permissions that are too wide.

File: Dockerfile [hunk 2]
```diff
@@ -1 +1,27 @@
...
+
+ENV TOMCAT_VERSION=9.0.22  \
+    TOMCAT_MAJOR_VERSION=9
+
+ADD  https://archive.apache.org/dist/tomcat/tomcat-${TOMCAT_MAJOR_VERSION}/v${TOMCAT_VERSION}/bin/apache-tomcat-${TOMCAT_VERSION}.tar.gz /tmp
+
+RUN cd /tmp && \
+    tar -xf apache-tomcat-${TOMCAT_VERSION}.tar.gz && \
+    rm apache-tomcat-${TOMCAT_VERSION}.tar.gz && \
+    mv apache-tomcat-${TOMCAT_VERSION} /opt/tomcat/ && \
```
Comment (fabricetriboix) [inline Dockerfile:18]: How is it possible that the `/opt/tomcat` directory exists at this stage? I don't understand...

File: Dockerfile [hunk 3]
```diff
@@ -1 +1,27 @@
...
+    chown -R tomcat: /opt/tomcat && \
```
Comment (fabricetriboix) [inline Dockerfile]: Same comment as before: All Tomcat files should be owned by `root:root` by default, and readable by the `tomcat` user. If Tomcat absolutely needs to be able to write to some files/directories, then only for those files/directories we can be less restrictive.

**Review by fabricetriboix — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 3 above)
Comment (fabricetriboix) [reply Dockerfile]: @naveen-armedia  Wherever Tomcat requires write access, you can provide it so. I am quite sure that Tomcat does not require write access for all files and directories under `/opt/tomcat`. For example the `README.md` won't be touched by Tomcat. This is a trivial example but hopefully you understand my point.

Please change the ownership of all files and directories under `/opt/tomcat` to `root:root`, and give write access to the `tomcat` user only for those files and directories that Tomcat absolutely need to write to. Thanks.

File: Dockerfile [hunk 4]
```diff
@@ -1 +1,28 @@
...
+ENV TOMCAT_VERSION=9.0.22  \
+    TOMCAT_MAJOR_VERSION=9
+
+ADD  https://archive.apache.org/dist/tomcat/tomcat-${TOMCAT_MAJOR_VERSION}/v${TOMCAT_VERSION}/bin/apache-tomcat-${TOMCAT_VERSION}.tar.gz /tmp
+
+RUN cd /tmp && \
+    tar -xf apache-tomcat-${TOMCAT_VERSION}.tar.gz && \
+    rm apache-tomcat-${TOMCAT_VERSION}.tar.gz && \
+    mv apache-tomcat-${TOMCAT_VERSION} /opt/tomcat/ && \
+    chown -R tomcat:tomcat /opt/tomcat && \
```
Comment (fabricetriboix) [inline Dockerfile:19]: Same comment as for the other PR

**Review by fabricetriboix — CHANGES_REQUESTED** [review]

**Comment by fabricetriboix** [comment]
@naveen-armedia In the current configuration, will all Tomcat logs to directed to stdout? No logs should be written anywhere inside the container. Thanks

File: Dockerfile (hunk 1 above)
Comment (naveen-armedia) [reply Dockerfile]: okay. changed it to tomcat group

**Review by fabricetriboix — DISMISSED** [review]

**Comment by ymuwakki** [comment]
@naveen-armedia  Please change the tomcat version to 9.0.50

File: Dockerfile [hunk 5]
```diff
@@ -1 +1,37 @@
+FROM 345280441424.dkr.ecr.ap-south-1.amazonaws.com/ark_base_java11:latest
+
+LABEL ORG="Armedia LLC" \
+      VERSION="1.0" \
+      IMAGE_SOURCE=https://github.com/ArkCase/ark_base_java11_tomcat9 \
+      MAINTAINER="Armedia LLC"
+
+RUN useradd  --system --user-group tomcat
+
+ENV TOMCAT_VERSION=9.0.22  \
```
Comment (ymuwakki) [inline Dockerfile]: @naveen-armedia  please change tomcat version to 9.0.50 (per @david-oc-miller )

File: Dockerfile (hunk 2 above)
Comment (JonHolman) [reply Dockerfile:18]: @naveen-armedia  is this resolved?

File: Dockerfile (hunk 4 above)
Comment (JonHolman) [reply Dockerfile:19]: @naveen-armedia  is this resolved?

File: Dockerfile [hunk 6]
```diff
@@ -1 +1,37 @@
...
+    rm apache-tomcat-${TOMCAT_VERSION}.tar.gz && \
+    mv apache-tomcat-${TOMCAT_VERSION} /opt/tomcat/ && \
+    chown -R tomcat:tomcat /opt/tomcat && \
+    chmod +x /opt/tomcat/bin/*.sh && \
+    # Removal of default/unwanted Applications
+    rm -f -r -d /opt/tomcat/webapps/* 
+
+ENV CATALINA_HOME /opt/tomcat/
+ENV CATALINA_PID /opt/tomcat/temp/tomcat.pid
+ENV PATH=/opt/tomcat/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```
Comment (JonHolman) [inline Dockerfile]: Can these be combined into one ENV statement to reduce the number of layers?

File: Dockerfile [hunk 7]
```diff
@@ -1 +1,37 @@
...
+
+LABEL ORG="Armedia LLC" \
+      VERSION="1.0" \
+      IMAGE_SOURCE=https://github.com/ArkCase/ark_base_java11_tomcat9 \
+      MAINTAINER="Armedia LLC"
+
+RUN useradd  --system --user-group tomcat
+
+ENV TOMCAT_VERSION=9.0.22  \
+    TOMCAT_MAJOR_VERSION=9
```
Comment (JonHolman) [inline Dockerfile]: Can the ENV lines below be combined with these?

File: Dockerfile (hunk 5 above)
Comment (naveen-armedia) [reply Dockerfile]: @ymuwakki  updated with 9.0.50 tomcat version

File: Dockerfile (hunk 7 above)
Comment (naveen-armedia) [reply Dockerfile]: Yeah updated with single ENV command.

File: Dockerfile (hunk 6 above)
Comment (naveen-armedia) [reply Dockerfile]: Yeah updated with single ENV command.

File: Dockerfile (hunk 2 above)
Comment (naveen-armedia) [reply Dockerfile:18]: @JonHolman the destination directory or file are  created if not present when using `mv` command.

**Review by JonHolman — APPROVED** [review]

**Review by fabricetriboix — APPROVED** [review]

**Pull request merged**